src/prefect/_internal/concurrency/timeouts.py

```diff
@@ -0,0 +1,254 @@
+import contextlib
+import ctypes
+import math
+import signal
+import sys
+import threading
+import time
+from typing import Optional, Type
+
+import anyio
+
+from prefect.logging import get_logger
+
+# TODO: We should update the format for this logger to include the current thread
+logger = get_logger("prefect._internal.concurrency.timeouts")
+
+
+class CancelContext:
+    """
+    Tracks if a cancel context manager was cancelled.
+
+    The `cancelled` property is threadsafe.
+    """
+
+    def __init__(self, timeout: Optional[float]) -> None:
+        self._timeout = timeout
+        self._cancelled: bool = False
+        self._lock = threading.Lock()
+
+    @property
+    def timeout(self) -> Optional[float]:
+        return self._timeout
+
+    @property
+    def cancelled(self):
+        with self._lock:
+            return self._cancelled
+
+    @cancelled.setter
+    def cancelled(self, value: bool):
+        with self._lock:
+            self._cancelled = value
```

Comment:
@chrisguidry do you know if this is necessary? I feel like you actually don't need a lock for threadsafe get/set of a single value and it'd only be needed if there was more going on in the `with`.

Reply:
I was puzzling over that but I thought you knew something I didn't HAHA.  I don't think you need a lock around either one of those.  It's definitely possible that they are multiple instructions under the hood, but that doesn't really matter, because the caller will either see one or the other value when they call it.  The real thing you would need a lock around is if you were trying to make a decision _with_ the value of the property:

```
with lock:
  if not thing.cancelled:
    thing.cancel()
```

Reply:
Makes sense; removed this.